certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -382,12 +382,28 @@ def choose_redirect_vhost(self, target_name, port):
 
         :param str target_name: domain name
         :param str port: port number
+        :param bool create_if_no_match: If we should create a new vhost from default
+            when there is no match found. If we can't choose a default, raise a
+            MisconfigurationError.
+
         :returns: vhost associated with name
         :rtype: :class:`~certbot_nginx.obj.VirtualHost`
 
         """
         matches = self._get_redirect_ranked_matches(target_name, port)
-        return self._select_best_name_match(matches)
+        vhost = self._select_best_name_match(matches)
+        if not vhost:
```

Comment:
This might just be temporary while you're still working on it, but I don't think we should do this here. Testing on a nginx config without a HTTP server block for a domain and requesting a redirect I got:

```
MisconfigurationError: Cannot find a VirtualHost matching domain example.org. In order for Certbot to correctly perform the challenge please add a corresponding server_name directive to your nginx configuration: https://nginx.org/en/docs/http/server_names.html
```

Reply:
We should do this. That was a bug because I was testing against an integer when something needed a string. Can you check if this is still happening?

Reply:
(This doesn't change the redirect behavior at all, it just reuses this method for http01.)

Reply:
(I'm assuming that failure was happening because it wasn't finding the server block to do a challenge with. I didn't catch that in my manual testing because I have a default block.)

Reply:
I can still reproduce. See https://github.com/certbot/certbot/blob/9e952081014b9545ce6ddb8b6ecc86a51bf94131/certbot-nginx/certbot_nginx/configurator.py#L600.

Reply:
Previously we were letting `choose_redirect_vhost` return `None` and checking that down below. Now we raise an exception out of `choose_redirect_vhost`.

Reply:
Ah cool that's easy enough to fix

Reply:
should be fixed now